## **Dynamo**: Mapping Vector Field of Single Cells
![Dynamo](https://www.dropbox.com/s/dvrfgo4qt5ispqs/dynamo_intro.png?raw=1)

Understanding how gene expression in single cells progress over time is vital for revealing the mechanisms governing cell fate transitions. RNA velocity, which infers immediate changes in gene expression by comparing levels of new (unspliced) versus mature (spliced) transcripts (La Manno et al. 2018), represents an important advance to these efforts. A key question remaining is whether it is possible to predict the most probable cell state backward or forward over arbitrary time-scales. To this end, we introduce an inclusive model (termed Dynamo) capable of predicting cell states over extended time periods, that incorporates promoter state switching, transcription, splicing, translation and RNA/protein degradation by taking advantage of scRNA-seq and the co-assay of transcriptome and proteome. We also implement scSLAM-seq by extending SLAM-seq to plate-based scRNA-seq (Hendriks et al. 2018; Erhard et al. 2019; Cao, Zhou, et al. 2019) and augment the model by explicitly incorporating the metabolic labelling of nascent RNA. We show that through careful design of labelling experiments and an efficient mathematical framework, the entire kinetic behavior of a cell from this model can be robustly and accurately inferred. Aided by the improved framework, we show that it is possible to analytically reconstruct the transcriptomic vector field from sparse and noisy vector samples generated by single cell experiments. The analytically reconstructed vector further enables global mapping of potential landscapes that reflects the relative stability of a given cell state, and the minimal transition time and most probable paths between any cell states in the state space This work thus foreshadows the possibility of predicting long-term trajectories of cells during a dynamic process instead of short time velocity estimates. Our methods are implemented as an open source tool, [dynamo](https://github.com/aristoteleo/dynamo-release).

## Dynamo workflow 

![dynamo_workflow](https://www.dropbox.com/s/awwcs97exu01kvs/dynamo_workflow.png?raw=1)

In the following, you can see python snippets that highlight each of the main steps in dynamo. 

Surely the first step in applying dynamo is to load some datasets of interests. Dynamo uses [AnnData](https://anndata.readthedocs.io/en/latest/) as its 
container so it you can load loom file, h5 or h5ad, etc. files via `read_loom`, `read_h5`, `read_h5ad`. Dynamo also provides a function (`read_NASC_seq`) to load output files 
from the [NASC-seq pipeline](https://github.com/sandberg-lab/NASC-seq) that applied to single cell based SLAM-seq experiments. You may also want to learn 
protein velocity, you should then attach the protein abundance data can to the `.obsm` attribute of the `AnnData` object. 

```python
# first step: Load data 
import dynamo as dyn 
# set matplotlib's rcParams. this setting tries to emulate ggplot style. 
dyn.configuration.set_figure_params('dynamo')  
# you can read your own data via 
adata = dyn.read_loom(filename, **param) # (or use read_h5ad, read_NASC_seq to load result generated from NASC-seq pipeline)
# here let us play with Dentate Gyrus example dataset
adata = dyn.sample_data.DentateGyrus()  # there are many sample datasets available. 
```

Next you may want to first check the fraction of the spliced, unspliced or new (metabolic labeled ) or total mRNAs in your 
data by `show_fraction`. Then you are ready to run the `recipe_monocle` function that performs preprocessing of the data in a single 
shot. `recipe_monocle` uses similar strategy from [Monocle 3](https://cole-trapnell-lab.github.io/monocle3/) to normalize all 
datasets in different layers (the spliced, unspliced or new (metabolic labeled ) or total mRNAs). You can check the `mean-dispersion` 
plot to see where do the automatically selected feature genes locate and also the `variance explained plot` to take a glance 
at the variance explained by principal components. 
```python
# second step: preprocess
dyn.pl.show_fraction(adata)
adata = dyn.pp.recipe_monocle(adata, n_top_genes=3000)
dyn.pl.featureGenes(adata)
dyn.pl.variance_explained(adata)
```

Next you will want to learn the velocity values for all genes that pass some filters (default it is all the selected feature genes) in each cell. 
the `dyn.tl.dynamics` do all the hard work for you. It automatically detects the data you have and learn the velocity vectors accordingly. For example, 
if you have scSLAM-seq data (identified by the `new` and `total` or `uu`, `ul`, `su`, `sl` layers), dynamo will learn the transcription, splicing and degradation 
rates for each gene. We developed two mode of estimation, either the `steady_state` assumption (from the seminal [RNA velocity](https://www.nature.com/articles/s41586-018-0414-6) work) 
or the `moment` model.  
```python
# third step: learning dynamics
dyn.tl.dynamics(adata)
```

You would like to also use different dimension reduction approach to reduce your scRNA-seq into low dimensional embedding. By default, we use the novel 
[trimap](https://github.com/eamid/trimap) dimension reduction method, which uses triplet constraints to form a low-dimensional embedding of a set of points. 
`trimap` is arguably more efficient than `UMAP` and better preserve the global structure of the data. Note that we also generalized diffusion map so we can 
performing dimension reduction via diffusion while also considering drift. A class of structural learning based dimension reduction methods from us will also 
be supported shortly.   
```python
# fourth step: reduce dimension 
dyn.tl.reduceDimension(adata, velocity_key='velocity_S',reduction_method='trimap')
```

You love to see the velocity vector on low dimensional embedding. To get there, you want to first apply our improved transition matrix reconstruction method. In
contrast to the "correlation kernel" from [velocyto](https://github.com/velocyto-team/velocyto.py) or [scVelo](https://github.com/theislab/scvelo), dynamo is powered 
by the [It√¥ kernel](https://twitter.com/Xiaojie_Qiu/status/1188875696178753537) that not only consider the correlation between the vector from any cell to its nearest 
neighbors and its velocity vector but also the correspondingly distance. We expect this new kernel will enable us to visualize more intricate vector flow or steady states 
in low dimension. We also expect it will improve the calculation of the stationary distribution or source states of sampled cells. 
```python
# fifth step: project velocity 
dyn.tl.cell_velocities(adata, vkey='pca', basis=basis, method='analytical', cores=1)
```

At this stage, you may already cannot wait to see how does the velocity vector looks like in low dimension space. Similar to velocyto and scvelo, we provide three plottinf utilities 
that visualize the cell-wise velocity vector, velocity vector on a space or the stream line plot that integrate along grid velocity vectors via fourth-order Runger-Kutta algorithm. 
Obviously we don't want to step here. Let us move to some real juice of dynamo in the next section to learn the velocity vector in the full transcriptomic space and map the potential 
landscape. 
```python
# sixth step: visualize vector field
dyn.pl.cell_wise_velocity(adata, genes=gene_list, basis=basis, n_columns=3) # ['GRIA3', 'LINC00982', 'AFF2']
dyn.pl.grid_velocity(adata, genes=gene_list, basis=basis, n_columns=2, figsize=[8, 8])  # ['GRIA3', 'LINC00982', 'AFF2']
dyn.pl.grid_velocity(adata, genes=gene_list,  basis=basis, n_columns=1, figsize=[8, 8], color=['ClusterName'])  # ['GRIA3', 'LINC00982', 'AFF2'] color=['ClusterName'],
dyn.pl.stremline_plot(adata, genes=gene_list, n_columns=2, figsize=[8, 8], density=3)  # ['GRIA3', 'LINC00982', 'AFF2'] color=['ClusterName']
```

Dynamo aspires to learn a vector field function in the transcriptomic space. With the learnt vector field, you can then predict the cell fate in high dimension over arbitrary time scale 
from any cell states. You can experiment it via the dyn.tl.VectorField or dyn.tl.fate function. 
```python
# seventh step: learn vector field
dyn.tl.VectorField(adata) 
dyn.tl.fate(adata)
dyn.pl.line_integral_conv(adata)
```

Potential landscape is an intuitive concept that used in various discipline. It provides a global description of cell state stability. Once we learnt vector field, dynamo allows to to map 
the potential landscape and the least action path that convert from cell type to another cell type. You can experiment it via the dyn.tl.Potential or dyn.tl.action function. 
```python
# eigth step: map potential landscape
dyn.tl.action(adata) 
dyn.pl.Potential(adata)
```

## Installation

Note that this is our first alpha version of **Dynamo** (as of July 9th, 2019). Dynamo is still under active development. Stable version of Dynamo will be released when it is ready. Until then, please use **Dynamo** with caution. We welcome any bugs reports (via GitHub issue reporter) and especially code contribution  (via GitHub pull requests) of **Dynamo** from users to make it an accessible, useful and extendable tool. For discussion about different usage cases, comments or suggestions related to our manuscript and questions regarding the underlying mathematical formulation of dynamo, we provided a google group [goolge group](https://groups.google.com/forum/#!forum/dynamo-user/). Dynamo developers can be reached by <xqiu.sc@gmail.com>. To install the newest version of dynamo, you can git clone our repo and then use::

```sh
pip install directory_to_dynamo_release_repo/
```

Alternatively, You can install **Dynamo** from source, using the following script:
```sh
pip install git+https://github.com:aristoteleo/dynamo-release
```

## Citation

Xiaojie Qiu, Yan Zhang, Dian Yang, Shayan Hosseinzadeh, Li Wang, Ruoshi Yuan, Song Xu, Yian Ma, Joseph Replogle, Spyros Darmanis, Jianhua Xing, Jonathan S Weissman (2019): Mapping Vector Field of Single Cells. BioRxiv

biorxiv link: https://www.biorxiv.org/content/10.1101/696724v1

## Theory behind dynamo
For the vector field reconstruction and potential landscape mapping, please refer to our preprint. We also released the complete derivation of the matrix form of the moment generation functions for parameter estimation in full_derivation.pdf file in the dynamo-notebook [GitHub repo](https://github.com/aristoteleo/dynamo-notebooks).

The dynamo-notebook repo also provides tutorials on how to use dynamo for reconstructing vector field, calculating least action path and potential of cell states.  

## Acknowledgement
We would like to sincerely thank the developers of velocyto (La Manno Gioele and others), scanpy (Alex Wolf and others) and svelo (Volker Bergen and others) on their amazing tools which demonstrate the best practice of scientific programming in Python. Dynamo takes various technical inspiration from those packages. Dynamo is (we are trying to and hope users can contribute to) fully compatible with those tools and velocity estimation from either scvelo or velocyto can both be used as input to learn the functional form of vector field for predicting the cell fate over extended time period and mapping global cell state potential. 

## Contribution 
If you want to contribute to the development of dynamo, please check out CONTRIBUTION instruction: [Contribution](https://github.com/aristoteleo/dynamo-release/blob/master/CONTRIBUTING.md)

## Documentation  
The documentation of dynamo package is available at [readthedocs](https://dynamo-release.readthedocs.io/en/latest/)
